compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2345,6 +2348,136 @@ public void visit(BLangMatchGuard matchGuard) {
         matchGuard.expr.accept(this);
     }
 
+    @Override
+    public void visit(BLangMappingMatchPattern mappingMatchPattern) {
+        BRecordTypeSymbol recordSymbol =
+                Symbols.createRecordSymbol(0, names.fromString("$anonRecordType$" + recordCount++),
+                        env.enclPkg.symbol.pkgID, null, env.scope.owner, mappingMatchPattern.pos, VIRTUAL);
+        LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+
+        for (BLangFieldMatchPattern fieldMatchPattern : mappingMatchPattern.fieldMatchPatterns) {
+            fieldMatchPattern.accept(this);
+            String fieldName = fieldMatchPattern.fieldName.value;
+            BVarSymbol fieldSymbol = new BVarSymbol(0, names.fromString(fieldName), env.enclPkg.symbol.pkgID,
+                    fieldMatchPattern.matchPattern.type, recordSymbol, fieldMatchPattern.pos, COMPILED_SOURCE);
+            BField field = new BField(names.fromString(fieldName), fieldMatchPattern.pos, fieldSymbol);
+            fields.put(fieldName, field);
+            mappingMatchPattern.declaredVars.putAll(fieldMatchPattern.declaredVars);
+        }
+        BRecordType recordVarType = new BRecordType(recordSymbol);
+        recordVarType.fields = fields;
+        recordVarType.restFieldType = symTable.anydataType;
+        if (mappingMatchPattern.restMatchPattern != null) {
```

Comment:
The rest match pattern needs to allow for all unmatched fields which can include non-rest fields in the case of a record.

The following doesn't fail atm.

```ballerina
public function main() {
    record {|
        int a;
        int b;
        string...;
    |} x = {a : 2, b: 3};

    match x {
        {a: 2, ...var rst} => {
            map<string> mp = rst; // no compilation error, even though the type of `rst`
                                  // is `record {|int b; string...;|}`  
        }
    }
}
```

Reply:
Fixed